File: src/main/kotlin/problems/array/sorting/BubbleSort.java
package problems.array.sorting;

import java.util.Arrays;

public class BubbleSort {

    public static int[] bubbleSort(int[] arr, int n) {
        boolean swapped;
        for (int i = 0; i < n - 1;i++) {
            swapped = false;
            for (int  j = 0; j < n - i - 1;j++) {

                if (arr[j+1] < arr[j]) {
                    swapped = true;
                    int temp = arr[j];
                    arr[j] = arr[j+1];
                    arr[j+1] = temp;
                }
            }

            if (!swapped) {
                break;
            }
        }

        return arr;
    }

    public static void bubbleSortRecursive(int[] arr, int i) {
        if (i == 1) return;
        for (int j = 0; j < i-2;j++) {
            if (arr[j] > arr[j+1]) {
                int temp = arr[j];
                arr[j] = arr[j+1];
                arr[j+1] = temp;
            }
        }
        bubbleSortRecursive(arr,i-1);
    }

    public static void main(String[] args) {
        int[] arr = new int[]{2, 13, 4, 1, 3, 6, 28};
        bubbleSortRecursive(arr,arr.length);
        System.out.println(Arrays.toString(arr));
    }
}

File: src/main/kotlin/problems/array/sorting/BubbleSort.kt
package problems.array.sorting


/**
* This algorithm selects the next adjacent element and compares it with the current element
* If the current element is greater than the next element, then it swaps the current element
* At each pass the largest element will be moved to the last of the array
* The inner loop run only in the unsorted part i.e (n-i-1) where n is the size of the array and i is the value of the outer loop
* */

fun bubbleSort(arr:IntArray) : IntArray {

    val n = arr.size

    for (i in 0 ..< n) {

        for (j in 0 ..< n-i-1) {
            if (arr[j] > arr[j+1]) {
                val temp = arr[j]
                arr[j] = arr[j+1]
                arr[j+1] = temp
            }
        }
    }

    return arr
}


fun main() {
    var arr = intArrayOf(64, 25, 12, 22, 11)
    arr = bubbleSort(arr)
    arr.forEach {
        print("${it}, ")
    }
}